src/dag/write.ts

```diff
@@ -0,0 +1,158 @@
+import {getRootAsMeta} from './meta.js';
+import type * as kv from '../kv/mod.js';
+import {HeadChange, toLittleEndian, fromLittleEndian} from './dag.js';
+import {chunkDataKey, chunkMetaKey, headKey, chunkRefCountKey} from './key.js';
+import {Read} from './read.js';
+import type {Chunk} from './chunk.js';
+import * as utf8 from '../utf8.js';
+
+export class Write {
+  private readonly _kvw: kv.Write;
+  private readonly _mutatedChunks = new Set<string>();
+  private readonly _changedHeads = new Map<string, HeadChange>();
+
+  constructor(kvw: kv.Write) {
+    this._kvw = kvw;
+  }
+
+  read(): Read {
+    return new Read(this._kvw);
+  }
+
+  async putChunk(c: Chunk): Promise<void> {
+    const {hash, data, meta} = c;
+    const key = chunkDataKey(hash);
+    const p1 = this._kvw.put(key, data);
+    let p2;
+    if (meta !== undefined) {
+      p2 = this._kvw.put(chunkMetaKey(hash), meta);
+    }
+    this._mutatedChunks.add(c.hash);
+    await p1;
+    await p2;
+  }
+
+  async setHead(name: string, hash: string | undefined): Promise<void> {
+    const oldHash = await this.read().getHead(name);
+    const hk = headKey(name);
+
+    let p1: Promise<void>;
+    if (hash === undefined) {
+      p1 = this._kvw.del(hk);
+    } else {
+      p1 = this._kvw.put(hk, utf8.encode(hash));
+    }
+
+    const v = this._changedHeads.get(name);
+    if (v === undefined) {
+      this._changedHeads.set(name, {new: hash, old: oldHash});
+    } else {
+      // Keep old if existing
+      v.new = hash;
+    }
+
+    await p1;
+  }
+
+  async commit(): Promise<void> {
+    await this.collectGarbage();
+    await this._kvw.commit();
+  }
+
+  async collectGarbage(): Promise<void> {
+    // We increment all the ref counts before we do all the decrements. This
+    // is so that we do not remove an item that goes from 1 -> 0 -> 1
+    const newHeads: (string | undefined)[] = [];
+    const oldHeads: (string | undefined)[] = [];
+    for (const changedHead of this._changedHeads.values()) {
+      oldHeads.push(changedHead.old);
+      newHeads.push(changedHead.new);
+    }
+
+    for (const n of newHeads) {
```

Comment:
Nothing in here seems to be validating that the tree is complete -- that there are no dangling hash references. I think that we should be doing that (and am surprised if the rust code didn't). You shouldn't be able to commit a dag write that has a dangling reference.

We should be able to find these during GC. It will be easier if we ensure that all chunks always have all three keys.